### PR TITLE
three simplified equations to catch all cases for gbmm flops

### PR DESCRIPTION
**Pull request by DSMishler**
As we discussed in our meeting on Tuesday. I did some further algebraic simplifications and I believe this converges on an optimal solution. Take a look when you get the chance!

### CODE REVIEW

**Review by luszczek — COMMENTED** [review]
What are the corner case examples for rectangular matrices?

**Comment by DSMishler** [comment]
First corner case: the matrix is very tall and its lower bandwidth `kl` does not account for this difference (`dim=6000x2000x2000,kl=999`). The bottom case would have subtracted an isosceles triangle whose height is correct but whose base would be too long (`5000` in this case, which is `3000` too large). The correction code instead subtracts a *trapezoid* off of the matrix, and that algebraic expression is simplified.

The second corner case is simply the transpose of the first corner case.

Now that I notice the algebra, I noticed I have distributed the wrong sign on the "-1.0" (an error small enough that I didn't notice it during testing, but when I was thinking about the problem geometrically earlier today. I'll push a fix!

File: test/flops.h [hunk 1]
```diff
@@ -103,11 +103,29 @@ static double  flops_sgemm(double m, double n, double k)
...
 //------------------------------------------------------------ gbmm
+// usually the bottom equation returned calculates the flops,
+// but some matrices are too long or too wide and require extra care
+// think rectangle minus trapezoid minus triangle and reduce:
+//        (m*k - (m-kl+m-k-kl-1.0)*0.5*k - (k-ku-1.0)*(k-ku)*0.5)*n;
+//        (m*k - (m-kl)*k+(k-1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
+//        (kl*k + (k-1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
 static double fmuls_gbmm(double m, double n, double k, double kl, double ku)
-    { return (m * k - (m - kl - 1.0) * (m - kl) * 0.5 - (k - ku - 1.0) * (k - ku) * 0.5) * n; }
+{
```
Comment (luszczek) [inline test/flops.h:113]: We don't do spaces at the end of line.

File: test/flops.h [hunk 2]
```diff
@@ -103,11 +103,29 @@ static double  flops_sgemm(double m, double n, double k)
...
+    if(m > k+kl)
+        return (kl*k + (k-1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
+    if(k > m+ku)
+        return (ku*m - (m-kl-1.0)*(m-kl)*0.5 + (m-1.0)*m*0.5)*n;
+    return (m*k - (m-kl-1.0)*(m-kl)*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
+}
 
 static double fadds_gbmm(double m, double n, double k, double kl, double ku)
-    { return (m * k - (m - kl - 1.0) * (m - kl) * 0.5 - (k - ku - 1.0) * (k - ku) * 0.5) * n; }
+{
```
Comment (luszczek) [inline test/flops.h:122]: We don't do spaces at the end of line.

File: test/flops.h [hunk 3]
```diff
@@ -112,18 +112,18 @@ static double  flops_sgemm(double m, double n, double k)
 static double fmuls_gbmm(double m, double n, double k, double kl, double ku)
 { 
     if(m > k+kl)
-        return (kl*k + (k-1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
+        return (kl*k + (k+1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
```
Comment (luszczek) [inline test/flops.h:115]: Formulas with `-` are more likely to be correct because they nicely deteriorate to `0` for special bandwidth sizes. This new formula loses this property.

File: test/flops.h [hunk 4]
```diff
@@ -112,18 +112,18 @@ static double  flops_sgemm(double m, double n, double k)
...
     if(k > m+ku)
-        return (ku*m - (m-kl-1.0)*(m-kl)*0.5 + (m-1.0)*m*0.5)*n;
+        return (ku*m - (m-kl-1.0)*(m-kl)*0.5 + (m+1.0)*m*0.5)*n;
```
Comment (luszczek) [inline test/flops.h:117]: Formulas with `-` are more likely to be correct because they nicely deteriorate to `0` for special bandwidth sizes. This new formula loses this property.

File: test/flops.h [hunk 5]
```diff
@@ -112,18 +112,18 @@ static double  flops_sgemm(double m, double n, double k)
...
     return (m*k - (m-kl-1.0)*(m-kl)*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
 }
 
 static double fadds_gbmm(double m, double n, double k, double kl, double ku)
 { 
     if(m > k+kl)
-        return (kl*k + (k-1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
+        return (kl*k + (k+1.0)*k*0.5 - (k-ku-1.0)*(k-ku)*0.5)*n;
```
Comment (luszczek) [inline test/flops.h:124]: Formulas with `-` are more likely to be correct because they nicely deteriorate to `0` for special bandwidth sizes. This new formula loses this property.

File: test/flops.h [hunk 6]
```diff
@@ -112,18 +112,18 @@ static double  flops_sgemm(double m, double n, double k)
...
     if(k > m+ku)
-        return (ku*m - (m-kl-1.0)*(m-kl)*0.5 + (m-1.0)*m*0.5)*n;
+        return (ku*m - (m-kl-1.0)*(m-kl)*0.5 + (m+1.0)*m*0.5)*n;
```
Comment (luszczek) [inline test/flops.h:126]: Formulas with `-` are more likely to be correct because they nicely deteriorate to `0` for special bandwidth sizes. This new formula loses this property.

File: test/flops.h (hunk 2 above)
Comment (DSMishler) [reply test/flops.h:122]: I didn't even notice this space was present. My apologies - I will remove it!

File: test/flops.h (hunk 6 above)
Comment (DSMishler) [reply test/flops.h:126]: Forgive me for failing to understand, but I don't see how the bandwidth size has anything to do with the `-`. This change was due to a failure to properly simplify the algebra on my first formula, which I fixed here. Whether the `-` was a `+` will have no effect on simplifying for specific bandwidths, for it is about whether or not `1` will be added or subtracted from `m` or `k`, neither of which relates to the bandwidth.

I will push the first change regarding the space and wait for your feedback. Could you help point towards what I'm missing?

Comment (luszczek) [reply test/flops.h:126]: I was simply talking about the fact that `m-1` becomes `0` when `m` is `1`. This is the correct result since there is nothing to do when `m==1`. There seems to be little foresight or safeguards in the initial formula and its fix. Just patching the patches and discussing the discussion. I asked for corner cases that could server as easy and verifiable assertions: with the triple of `m,n,k==7,5,3` there is a manageable small set of `kl,ku` pairs to verify. The general guideline is that the code that goes into the mathematical library is a fraction of the code used for testing and is left out. PR discussion is best left for design, form, or cross-cutting issues but not for correctness.

**Comment by luszczek** [comment]
Simple examples to show mainline and corner cases would help in review. It's not included in the commit comments nor in the discussion. And the fixes in the subsequent commits seem to be going back-and-forth on some fundamental formulas. I'm not sure what to review and what to check.

File: test/flops.h (hunk 6 above)
Comment (DSMishler) [reply test/flops.h:126]: In the third expression which you are commenting about, in the corner case when `m==1`: there *is* work to be done! The first expression accounts for the terms in the band, the second expression (which, like I said, was incorrect the first time) needs to account for the term in the diagonal. It would draw a triangle for a `m>1` case, but for `m=1` this is only a single element. The third term (which would subtract a triangle) *will* have this property of nicely deteriorating to zero. Just like the mainline equation case in LAWN41, my equation case will have exactly one of its terms nicely deteriorate to zero. I've attached a picture to explain my point. [The case I was talking about.](https://docs.google.com/presentation/d/1b8DSjmpInchSwy9-eGnyXlhHgkhL2TCVAp8jmQRwupw/edit?usp=sharing)

LAWN41-GBMM-explain-m-1


What I *do* agree with, however, is that the form of this equation could perhaps be improved. Look at that rectangle in the equation. Notice that it happens to always be able to be one element larger. It could absorb this element, make the triangle smaller, and then reclaim the property of nicely deteriorating to zero. Of course, in this case, we would also lose out on a deterioration for `kl=0`. But why stop there: we could just combined the first two terms into the calculation of a trapezoid. I do trust your judgement on this call, as your intuition on what properties equations should have is surely more developed than mine ([m=3 case](https://docs.google.com/presentation/d/1b8DSjmpInchSwy9-eGnyXlhHgkhL2TCVAp8jmQRwupw/edit?usp=sharing)).

LAWN41-GBMM-explain-m-3


I do agree that test cases would be helpful for review, so I went ahead and performed several and attached them to this thread so you can review them if it's helpful. As for discussing correctness in a PR, I'm conflicted. I would like to point out that you also committed an equation that produced incorrect results for flops (to be precise, this equation was added to PLASMA as one of my PRs was closed, a PR whose code *did* produce the correct answer for flops for rectangular matrices). Furthermore, the author of the equation in the BLAS apparently did not test for the case of a rectangular band matrix either. If discussing correctness here is not appropriate, I will happily comply with another method you suggest.

In the general case, however, I agree and in the future will be sure that questions on the correctness of my code are answered well before the PR stage begins.

Finally, forgive me for asking for clarification, but what does "The general guideline is that the code that goes into the mathematical library is a fraction of the code used for testing and is left out" mean? We leave out the code that goes into the mathematical library? This does not seem right.

simple test cases: [gbmm_flops_tester.zip](https://github.com/icl-utk-edu/plasma/files/7705247/gbmm_flops_tester.zip)

**Comment by DSMishler** [comment]
Simple examples provided above in test case. Hope this is helpful. The current set of equations: as-is. Review those. If you request me to start another PR with a single commit, say the word and I'll comply.

**Comment by luszczek** [comment]
Wow, what a treasure trove of cases! Thank you for putting it together. It will be a great resource for understanding what was missing from LAWN41.
